### PR TITLE
fix: update TypeScript to v5

### PR DESCRIPTION
**Pull request by tychota**
# What problem am I fixing

Recently, Typescript 5.0 was released.

Type checking decorator used to be buggy [1], and that was fixed in part of the "Implement Stage3 Decorator proposal" [2].
Sadly, it means that the type checking is now more strict, and new error pops.

On a real project, that means:

image

NestJS fixed it in version 9.3.0 (see issue [3] and commit [4]).

This PR upgrade nest dependencies.

#  What is done

The work is done in 4 atomics commit. If you wich, you can safely remove any one of those before merging or cherry pick them on another PR if it is clearer.

- first commit (79eba68227e53b3caae717b0ba0e419c3e3205d3) backports changes by upgrading dev dependencies
- second commit (196f85847470899c71c9da7e60502b550608a3af) changes typescript version. As https://github.com/mikro-orm/mikro-orm/commit/ced7c9757187d999448ecad8a12051c52f5c8097 already upgraded main repo, it looked safe to me to do that here as well. Feel free to postpone this by removing the commit
- third (708985bd6b4dbb5dbbd9bd1c9061f9892231a703) and last commit (46e91a98436c0dd4a965797bd22df27e1a8e1f00) upgrade contributing.md with accurate docs. Feel free to reword, I'm not native

# Test plan

- build works and include the proper signature

image

# Conclusing

Let me know if i can do other fixes and thank you for the great works on mikro orm.

# Links

[1] https://github.com/microsoft/TypeScript/issues/52435#issuecomment-1405483291

[2] https://github.com/microsoft/TypeScript/pull/50820

[3] https://github.com/nestjs/nest/issues/10959 

[4] https://github.com/nestjs/nest/commit/268d9306d48720c68f6a16d4cb434c9176fe0a0f

### CODE REVIEW

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -41,6 +41,9 @@ Before you submit your pull request consider the following guidelines:
 - Search [GitHub](https://github.com/mikro-orm/mikro-orm/pulls) for an open or closed PR
   that relates to your submission. You don't want to duplicate effort.
 
+- Ensure that you are using yarn 3.x and not npm or yarn 1.x. This way, if you upgrade any dependency, 
+  the lockfile will have the same format as the one used by the project.
```
Comment (B4nan) [inline CONTRIBUTING.md:45]:
```suggestion
- Fork the project and install NPM dependencies. This project uses `yarn/berry`, so you will need to install it, preferably by [enabling `corepack`](https://yarnpkg.com/getting-started/install).
```

this should be aligned with the contributing guide from the ORM repo - all you need to care about is enabling corepack, you don't have to manage the versions yourself

File: tsconfig.build.json [hunk 2]
```diff
@@ -7,15 +7,13 @@
...
     "experimentalDecorators": true,
     "emitDecoratorMetadata": true,
     "outDir": "./dist"
   },
   "include": ["src/**/*"],
-  "exclude": [
-    "**/node_modules",
-    "**/dist"
-  ]
+  "exclude": ["**/node_modules", "**/dist"]
```
Comment (B4nan) [inline tsconfig.build.json:18]: unrelated change

```suggestion
  "exclude": [
    "**/node_modules",
    "**/dist"
  ]
```

File: tsconfig.build.json [hunk 3]
```diff
@@ -7,6 +7,7 @@
     "declaration": true,
     "sourceMap": false,
     "strict": true,
+    "ignoreDeprecations": "5.0",
```
Comment (B4nan) [inline tsconfig.build.json:10]: better to fix this instead of ignoring it. it should be pretty simple given the project scope.

**Review by B4nan — APPROVED** [review]
Thanks!

File: tsconfig.build.json (hunk 3 above)
Comment (B4nan) [reply tsconfig.build.json:10]: Have you actually tried to build the project before you added this? It looks like there are no issues without it.

**Comment by tychota** [comment]
Thank for merging it.
